project-chip/connectedhomeip
# [WIP] [DO NOT SUBMIT] Device to device demo

#### Problem

This PR works as a demonstration on how we can use https://github.com/project-chip/connectedhomeip/pull/9103 and https://github.com/project-chip/connectedhomeip/pull/9200 to setup a device to device CASE session and send cluster commands.

#### Change overview

The change shows how to extend `InteractionPeerDelegate` for device to device interaction.
The change contains a tailored version of `CHIPClusters.cpp` copied from `src/controller/data_model/zap-generated/`.

For the final device to device solution we need to work out a way to make the generated cluster server and client code coexist. I didn't have time to study the code generation scripts and templates. I hacked the generated code to get the basic on/off toggling working.

#### Testing

Usage:
* Build linux all-cluster-app with `gn gen out/debug --args='chip_build_libshell=true' && ninja -C out/debug`
* Pair linux all-cluster-app with `chip-tool pairing ethernet ${PAIRING_CODE} ${DISCRIMINATOR} ${IP_ADDR} 5540`
* Pair another common light-app or all-cluster-app device with chip-tool and make sure the device is under the same LAN as the linux all-cluster-app. (Personally I'm using docker to launch two all-cluster-app in different containers.)
* Fetch the node ID and the IP address of the device to be controlled with `avahi-browse -r _matter._tcp -t`
* Run the following commands in the linux all-cluster-app
```
> control connect ${DEVICE_IP} ${DEVICE_NODE_ID}
> control toggle
```
